File: Data Structure/src/python/sorting/merge_sort.py
def merge(arr, low, mid, high):
    temp = []
    left = low
    right = mid + 1
    # Sort and merge
    while left <= mid and right <= high:
        if arr[left] <= arr[right]:
            temp.append(arr[left])
            left += 1
        else:
            temp.append(arr[right])
            right += 1

    # No comparison as they are sorted.
    while left <= mid:
        temp.append(arr[left])
        left += 1

    while right <= high:
        temp.append(arr[right])
        right += 1

    for i in range(low, high + 1):
        arr[i] = temp[i - low]  # Because temp is 0-indexed


def merge_sort(arr, low, high):
    if low >= high:
        return

    mid = (low + high) // 2
    merge_sort(arr, low, mid)  # Left half
    merge_sort(arr, mid + 1, high)  # Right half
    merge(arr, low, mid, high)


if __name__ == "__main__":
    arr = [64, 34, 25, 12, 22, 11, 90]
    print("Unsorted array:", arr)
    merge_sort(arr, 0, len(arr) - 1)
    print("Sorted array:", arr)

    arr = [i for i in range(10)]
    print("Unsorted array:", arr)
    merge_sort(arr, 0, len(arr) - 1)
    print("Sorted array:", arr)

File: Data Structure/src/go/sorting/merge_sort.go
package main

import "fmt"

func merge(arr []int, low int, mid int, high int) {
	var left int = low
	var right int = mid + 1
	var temp []int

	for left <= mid && right <= high {
		if arr[left] <= arr[right] {
			temp = append(temp, arr[left])
			left++
		} else {
			temp = append(temp, arr[right])
			right++
		}
	}

	for left <= mid {
		temp = append(temp, arr[left])
		left++
	}

	for right <= high {
		temp = append(temp, arr[right])
		right++
	}

	for i := low; i <= high; i++ {
		arr[i] = temp[i-low]
	}

}

func merge_sort(arr []int, low int, high int) {
	if low >= high {
		return
	}
	var mid int = (low + high) / 2
	merge_sort(arr, low, mid)
	merge_sort(arr, mid+1, high)
	merge(arr, low, mid, high)
}

func main() {
	arr := []int{64, 34, 25, 12, 22, 11, 90, 78}
	var high int = len(arr)
	var low int = 0
	merge_sort(arr, low, high-1)
	fmt.Println("Sorted Array: ", arr)
}
